File: src/graphics/cpp/moving_balls/movable_image.cpp
#include "movable_image.h"
#include <QGraphicsScene>
#include <QPainter>
#include <cstdlib>

const int rectSize = 100;

MovableImage::MovableImage(QRectF rectBorder, QObject *parent)
    : QObject(parent), QGraphicsItem() {

  // random between -5 and 5
  auto speedX = (std::rand() % 10) - 5;
  auto speedY = (std::rand() % 10) - 5;
  speed = QPointF(speedX, speedY);

  // random start position
  auto startX =
      (std::rand() % (int)(rectBorder.width() - rectSize)) + rectBorder.x();
  auto startY =
      (std::rand() % (int)(rectBorder.height() - rectSize)) + rectBorder.y();

  QPointF point(startX, startY);
  setPos(point);

  // make selectable and movable
  setFlag(QGraphicsItem::ItemIsSelectable);
  setFlag(QGraphicsItem::ItemIsMovable);
}

QRectF MovableImage::boundingRect() const { return QRect(0, 0, 20, 20); }

void MovableImage::paint(QPainter *painter,
                         const QStyleOptionGraphicsItem *option,
                         QWidget *widget) {
  // gl shade model for smooth edges
  painter->setRenderHint(QPainter::SmoothPixmapTransform);
  auto rect = boundingRect();
  QBrush brush = painter->brush();

  // collision detection
  if (scene()->collidingItems(this).isEmpty()) {
    QPen pen(Qt::green, 5);
    brush = QBrush(Qt::green);
    painter->setPen(pen);
  } else {
    QPen pen(Qt::red, 5);
    painter->setPen(pen);
    brush = QBrush(Qt::red);
    collideWithWall();
  }

  painter->setBrush(brush);
  painter->drawEllipse(rect);
}

void MovableImage::advance(int phase) {
  setPos(x() + speed.x(), y() + speed.y());
}

void MovableImage::collideWithWall() {
  auto x = pos().x();
  auto y = pos().y();

  // get the position of the walls
  auto sceneRect = scene()->sceneRect();
  auto leftWall = sceneRect.x();
  auto rightWall = sceneRect.x() + sceneRect.width();
  auto topWall = sceneRect.y();
  auto bottomWall = sceneRect.y() + sceneRect.height();

  // if the ball hits the left wall
  if (x <= leftWall) {
    // bounce the ball off the wall
    speed = QPointF(-speed.x(), speed.y());
    x = leftWall + 1;
  }

  // if the ball hits the right wall
  if (x + boundingRect().width() >= rightWall) {
    // bounce the ball off the wall
    speed = QPointF(-speed.x(), speed.y());
    x = rightWall - boundingRect().width() - 1;
  }

  // if the ball hits the top wall
  if (y <= topWall) {
    // bounce the ball off the wall
    speed = QPointF(speed.x(), -speed.y());
    y = topWall + 1;
  }

  // if the ball hits the bottom wall
  if (y + boundingRect().height() >= bottomWall) {
    // bounce the ball off the wall
    speed = QPointF(speed.x(), -speed.y());
    y = bottomWall - boundingRect().height() - 1;
  }

  // set the position
  setPos(x, y);
}

QPointF MovableImage::getSpeed() { return speed; }

void MovableImage::setSpeed(QPointF newSpeed) { speed = newSpeed; }

void MovableImage::reverseSpeed() { speed = QPointF(-speed.x(), -speed.y()); }

File: src/graphics/python/moving_balls/movable_image.py
from random import randint, uniform

from PyQt6.QtCore import QPointF, QRectF, QObject, QPoint
from PyQt6.QtGui import QBrush, QPen, QColor, QPainter
from PyQt6.QtWidgets import QGraphicsItem

rectSize = 30


class MovableImage(QGraphicsItem, QObject):
    def __init__(self, rectBorder, parent=None):
        super(QGraphicsItem, self).__init__()
        # super(QObject, self).__init__(parent)
        self.setFlag(QGraphicsItem.GraphicsItemFlag.ItemIsMovable)
        self.setFlag(QGraphicsItem.GraphicsItemFlag.ItemIsSelectable)
        self.speed = self.randomize_speed()
        start_x = (randint(0, rectBorder.width() - rectSize)) + rectBorder.x()
        start_y = (randint(0, rectBorder.height() - rectSize)) + rectBorder.y()
        self.setPos(start_x, start_y)

    def boundingRect(self):
        return QRectF(-rectSize / 2, -rectSize / 2, rectSize, rectSize)

    def paint(self, painter, option, widget):
        painter.setRenderHint(QPainter.RenderHint.SmoothPixmapTransform)
        rect = self.boundingRect()

        if len(self.scene().collidingItems(self)) > 0:
            pen = QPen(QColor("red"))
            painter.setPen(pen)
            brush = QBrush(QColor("red"))
            self.collide_wall()
        else:
            pen = QPen(QColor("green"))
            brush = QBrush(QColor("green"))
            painter.setPen(pen)
        painter.setBrush(brush)
        painter.drawEllipse(rect)

    def advance(self, phase):
        self.setPos(self.x() + self.speed.x(), self.y() + self.speed.y())

    def collide_wall(self):
        x = self.pos().x()
        y = self.pos().y()
        scene_rect = self.scene().sceneRect()
        left_wall = scene_rect.x()
        right_wall = scene_rect.x() + scene_rect.width()
        top_wall = scene_rect.y()
        bottom_wall = scene_rect.y() + scene_rect.height()
        if x - self.boundingRect().width() <= left_wall:
            self.speed = QPointF(abs(self.speed.x()), self.speed.y())
            x = left_wall + self.boundingRect().width()
        if x + self.boundingRect().width() >= right_wall:
            self.speed = QPointF(-abs(self.speed.x()), self.speed.y())
            x = right_wall - self.boundingRect().width()
        if y - self.boundingRect().height() <= top_wall:
            self.speed = QPointF(self.speed.x(), abs(self.speed.y()))
            y = top_wall + self.boundingRect().height()
        if y + self.boundingRect().height() >= bottom_wall:
            self.speed = QPointF(self.speed.x(), -abs(self.speed.y()))
            y = bottom_wall - self.boundingRect().height()
        self.setPos(x, y)

    def randomize_speed(self):
        speed_x = uniform(-5, 5)
        speed_y = uniform(-5, 5)
        self.speed = QPointF(speed_x, speed_y)
        return self.speed

    def reverse_speed(self):
        self.speed = QPointF(-self.speed.x(), -self.speed.y())

    def speed_direction(self):
        try:
            x_dir = self.speed.x() // abs(self.speed.x())
            y_dir = self.speed.y() // abs(self.speed.y())
        except ZeroDivisionError:
            self.randomize_speed()
            return self.speed_direction()

        return QPoint(x_dir, y_dir)
